src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/TypeFixupSignature.cs

```diff
@@ -39,11 +42,114 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
                 EcmaModule targetModule = factory.SignatureContext.GetTargetModule(_typeDesc);
                 SignatureContext innerContext = dataBuilder.EmitFixup(factory, _fixupKind, targetModule, factory.SignatureContext);
                 dataBuilder.EmitTypeSignature(_typeDesc, innerContext);
+
+                if (_fixupKind == ReadyToRunFixupKind.Check_TypeLayout)
+                {
+                    EncodeTypeLayout(dataBuilder, _typeDesc);
+                }
             }
 
             return dataBuilder.ToObjectData();
         }
 
+        private static void EncodeTypeLayout(ObjectDataSignatureBuilder dataBuilder, TypeDesc type)
+        {
+            Debug.Assert(type.IsValueType);
+            MetadataType defType = (MetadataType)type;
+
+            int pointerSize = type.Context.Target.PointerSize;
+            int size = defType.InstanceFieldSize.AsInt;
+            int alignment = GetClassAlignmentRequirement(defType);
+            ReadyToRunTypeLayoutFlags flags = ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_Alignment | ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_GCLayout;
+            if (alignment == pointerSize)
+            {
+                flags |= ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_Alignment_Native;
+            }
+
+            if (!defType.ContainsGCPointers)
+            {
+                flags |= ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_GCLayout_Empty;
+            }
+
+            bool isHfa = defType.IsHfa && defType.Context.Target.Architecture == TargetArchitecture.ARM || defType.Context.Target.Architecture == TargetArchitecture.ARM64;
+            if (isHfa)
+            {
+                flags |= ReadyToRunTypeLayoutFlags.READYTORUN_LAYOUT_HFA;
+            }
+
+            dataBuilder.EmitUInt((uint)flags);
+            dataBuilder.EmitUInt((uint)size);
+
+            if (isHfa)
+            {
+                switch (defType.HfaElementType.Category)
+                {
+                    case TypeFlags.Single:
+                        dataBuilder.EmitUInt((uint)CorElementType.ELEMENT_TYPE_R4);
+                        break;
+                    case TypeFlags.Double:
+                        dataBuilder.EmitUInt((uint)CorElementType.ELEMENT_TYPE_R4);
+                        break;
+                }
+            }
+            
+            if (alignment != pointerSize)
+            {
+                dataBuilder.EmitUInt((uint)alignment);
+            }
+
+            if (defType.ContainsGCPointers)
+            {
+                // Encode the GC pointer map
+                GCPointerMap gcMap = GCPointerMap.FromInstanceLayout(defType);
+
+                byte[] encodedGCRefMap = new byte[(size / pointerSize + 7) / 8];
```

Comment:
Is <code>size</code> always guaranteed to be a multiple of <code>pointerSize</code>? If not, do we need another upwards rounding in the division?

Reply:
`size` is not always a multiple of `pointerSize`, but my understanding is we only care about the number of full pointers that can fit into the type for the purposes of the GC ref map. This implementation matches the runtime's `CEECompileInfo::EncodeTypeLayout` implementation so I believe we want to keep them in sync.

Reply:
I see, that makes sense, thanks for clarifying.